firedrake/assemble.py

```diff
@@ -450,65 +555,86 @@ def apply_bcs(tensor, bcs, *, assembly_rank=None, form_compiler_parameters=None,
                     # ComponentFunctionSpace, check parent index
                     if V.parent.index is not None:
                         # Mixed, index doesn't match
                         if V.parent.index != i:
                             continue
                         # Index matches
-                    yield functools.partial(op2tensor[i, j].set_local_diagonal_entries, nodes, idx=V.component)
+                    op2tensor[i, j].set_local_diagonal_entries(nodes, idx=V.component)
                 elif V.index is None:
-                    yield functools.partial(op2tensor[i, j].set_local_diagonal_entries, nodes)
+                    op2tensor[i, j].set_local_diagonal_entries(nodes)
                 else:
                     raise RuntimeError("Unhandled BC case")
-        for bc in equationbcs:
-            yield from _assemble(bc.f, tensor=tensor, bcs=bc.bcs,
-                                 form_compiler_parameters=form_compiler_parameters,
-                                 mat_type=mat_type,
-                                 sub_mat_type=sub_mat_type,
-                                 appctx=appctx,
-                                 assemble_now=assemble_now,
-                                 zero_tensor=False)
-    elif assembly_rank == AssemblyRank.VECTOR:
-        for bc in dirichletbcs:
-            if assemble_now:
-                if diagonal:
-                    yield functools.partial(bc.set, tensor, 1)
+    elif assembly_rank == _AssemblyRank.VECTOR:
+        for bc in bcs:
+            if opts.assembly_type == _AssemblyType.SOLUTION:
+                if opts.diagonal:
+                    bc.set(tensor, 1)
                 else:
-                    yield functools.partial(bc.apply, tensor)
+                    bc.apply(tensor)
+            elif opts.assembly_type == _AssemblyType.RESIDUAL:
+                bc.zero(tensor)
             else:
-                yield functools.partial(bc.zero, tensor)
-        for bc in equationbcs:
-            if diagonal:
-                raise NotImplementedError("diagonal assembly and EquationBC not supported")
-            yield functools.partial(bc.zero, tensor)
-            yield from _assemble(bc.f, tensor=tensor, bcs=bc.bcs,
-                                 form_compiler_parameters=form_compiler_parameters,
-                                 mat_type=mat_type,
-                                 sub_mat_type=sub_mat_type,
-                                 appctx=appctx,
-                                 assemble_now=assemble_now,
-                                 zero_tensor=False)
+                raise AssertionError
+    elif assembly_rank == _AssemblyRank.SCALAR:
+        pass
+    else:
+        raise AssertionError
+
+
+@utils.known_pyop2_safe
+def _make_parloops(expr, tensor, bcs, diagonal, fc_params, assembly_rank):
+    """Create some parloops.
+
+    :arg expr: The expression to be assembled.
+    :arg tensor: The tensor to write to.
```

Comment:
Add - ```will be either a scalar, vector (masquerading as a :class:`.Function`) or :class:`.Matrix` depending on the other arguments given``` or similar. I presume it's actually decided by some combination of `assembly_rank` and `diagonal`?